test/database/openshiftclusters.go

```diff
@@ -0,0 +1,173 @@
+	"fmt"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/fvbommel/sortorder"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	"github.com/Azure/ARO-RP/pkg/database"
+	"github.com/Azure/ARO-RP/pkg/database/cosmosdb"
+)
+
+type ByKey []*api.OpenShiftClusterDocument
+
+func (a ByKey) Len() int           { return len(a) }
+func (a ByKey) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a ByKey) Less(i, j int) bool { return sortorder.NaturalLess(a[i].Key, a[j].Key) }
+
+func getQueuedOpenShiftDocuments(client cosmosdb.OpenShiftClusterDocumentClient) (res []*api.OpenShiftClusterDocument) {
+	docs := fakeOpenShiftClustersGetAllDocuments(client)
+
+	for _, r := range docs {
+		var include bool
+
+		switch r.OpenShiftCluster.Properties.ProvisioningState {
+		case
+			api.ProvisioningStateCreating,
+			api.ProvisioningStateUpdating,
+			api.ProvisioningStateAdminUpdating,
+			api.ProvisioningStateDeleting:
+			include = true
+		}
+
+		if include && (r.LeaseExpires > 0 && int64(r.LeaseExpires) < time.Now().Unix()) {
+			include = false
+		}
+
+		if include {
+			res = append(res, r)
+		}
+	}
+	return
+}
+
+func fakeOpenShiftClustersQueueLengthQuery(client cosmosdb.OpenShiftClusterDocumentClient, query *cosmosdb.Query, options *cosmosdb.Options) cosmosdb.OpenShiftClusterDocumentRawIterator {
+	results := getQueuedOpenShiftDocuments(client)
+	return &fakeOpenShiftClustersQueueLengthIterator{resultCount: len(results)}
+}
+
+func fakeOpenShiftClustersDequeueQuery(client cosmosdb.OpenShiftClusterDocumentClient, query *cosmosdb.Query, options *cosmosdb.Options) cosmosdb.OpenShiftClusterDocumentRawIterator {
+	docs := getQueuedOpenShiftDocuments(client)
+	return cosmosdb.NewFakeOpenShiftClusterDocumentClientRawIterator(docs, 0)
+}
+
+func fakeOpenshiftClustersMatchQuery(client cosmosdb.OpenShiftClusterDocumentClient, query *cosmosdb.Query, options *cosmosdb.Options) cosmosdb.OpenShiftClusterDocumentRawIterator {
+	var results []*api.OpenShiftClusterDocument
+
+	startingIndex, err := fakeOpenShiftClustersGetContinuation(options)
+	if err != nil {
+		return cosmosdb.NewFakeOpenShiftClusterDocumentClientErroringRawIterator(err)
+	}
+
+	docs := fakeOpenShiftClustersGetAllDocuments(client)
+	for _, r := range docs {
+		var key string
+		switch query.Parameters[0].Name {
+		case "@key":
+			key = r.Key
+		case "@clientID":
+			key = r.ClientIDKey
+		case "@resourceGroupID":
+			key = r.ClusterResourceGroupIDKey
+		default:
+			return cosmosdb.NewFakeOpenShiftClusterDocumentClientErroringRawIterator(cosmosdb.ErrNotImplemented)
+		}
+		if key == query.Parameters[0].Value {
+			results = append(results, r)
+		}
+	}
+	return cosmosdb.NewFakeOpenShiftClusterDocumentClientRawIterator(results, int(startingIndex))
+}
+
+func fakeOpenShiftClustersGetAllDocuments(client cosmosdb.OpenShiftClusterDocumentClient) []*api.OpenShiftClusterDocument {
+	input, err := client.ListAll(context.Background(), nil)
+	if err != nil {
+		// TODO: should this never happen?
+		panic(err)
```

Comment:
should the query signature in cosmosdb include an error return?

Reply:
no - return a magic error iterator that returns the error when Next() is called

Reply:
done